### PR TITLE
 User inputs migration

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
contains https://github.com/Drakkar-Software/OctoBot-Trading/pull/736

### CODE REVIEW

File: octobot_trading/exchange_data/exchange_symbol_data.py [hunk 1]
```diff
@@ -83,14 +90,30 @@ def handle_liquidations(self, liquidations):
         self.recent_trades_manager.add_new_liquidations(liquidations)
 
     def handle_mark_price_update(self, mark_price, mark_price_source) -> bool:
+        trigger_init_event = not self.prices_manager.initialized()
         updated = self.prices_manager.set_mark_price(mark_price, mark_price_source)
         if updated:
+            if trigger_init_event:
+                self._set_initialized_event(commons_enums.InitializationEventExchangeTopics.PRICE.value)
```
Comment (Herklos) [inline octobot_trading/exchange_data/exchange_symbol_data.py:97]: 👍 We can remove `price_events_manager.py ` now

File: octobot_trading/exchanges/exchange_manager.py [hunk 2]
```diff
@@ -81,29 +82,36 @@ def __init__(self, config, exchange_class_string):
     async def initialize_impl(self):
         await exchanges.create_exchanges(self)
 
-    async def stop(self, warning_on_missing_elements=True):
+    async def stop(self, warning_on_missing_elements=True, enable_logs=True):
```
Comment (Herklos) [inline octobot_trading/exchanges/exchange_manager.py:85]: I don't really like this approach, why not using debug level?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 3]
```diff
@@ -16,32 +16,53 @@
...
 
 class AbstractTradingMode(abstract_tentacle.AbstractTentacle):
     __metaclass__ = abc.ABCMeta
+    USER_INPUT_TENTACLE_TYPE = common_enums.UserInputTentacleTypes.TRADING_MODE
 
     MODE_PRODUCER_CLASSES = []
     MODE_CONSUMER_CLASSES = []
     # maximum seconds before sending a trading signal if orders are slow to create on exchange
     TRADING_SIGNAL_TIMEOUT = 10
+    SAVED_RUN_METADATA_DB_BY_BOT_ID = {}
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:65]: I don't understand why we need this static attribute here (instead of having a class attribute)

File: octobot_trading/modes/abstract_trading_mode.py [hunk 4]
```diff
@@ -80,6 +101,10 @@ def __init__(self, config, exchange_manager):
         # True when this trading mode is waken up only after full candles close
         self.is_triggered_after_candle_close = False
 
+        self.start_time = time.time()
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:104]: Why not in `start()` ?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 5]
```diff
@@ -118,6 +143,23 @@ def get_supported_exchange_types(cls) -> list:
...
+    def get_mode_producer_classes(self) -> list:
+        return self.MODE_PRODUCER_CLASSES
+
+    def get_mode_consumer_classes(self) -> list:
+        return self.MODE_CONSUMER_CLASSES
+
+    def should_emit_trading_signals_user_input(self, inputs: dict):
+        if self.user_input(
+            common_constants.CONFIG_EMIT_TRADING_SIGNALS, common_enums.UserInputTypes.BOOLEAN, False, inputs,
+            title="Emit trading signals on Astrolab for people to follow.", order=998
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:155]: What is `order=998`?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 6]
```diff
@@ -211,10 +270,107 @@ async def create_consumers(self) -> list:
...
+            user_commands_consumer = \
+                await channels.get_chan(services_channels.UserCommandsChannel.get_name()).new_consumer(
+                    self.user_commands_callback,
+                    {"bot_id": self.bot_id, "subject": self.get_name()}
+                )
+            base_consumers.append(user_commands_consumer)
+        except ImportError:
+            self.logger.warning("Can't connect to services channels")
+        except KeyError:
+            self.logger.debug(f"{services_channels.UserCommandsChannel.get_name()} unavailable")
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:288]: Can we put this code in a new method named `create_user_input_consumer` please?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 7]
```diff
@@ -211,10 +270,107 @@ async def create_consumers(self) -> list:
...
+        return base_consumers + await self._add_temp_consumers()
+
+    # TODO remove when proper run storage strategy
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:291]: Should we discuss it before merging?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 8]
```diff
@@ -244,6 +400,40 @@ def load_config(self) -> None:
...
             self.set_default_config()
 
+    LIGHT_VOLUME_WEIGHT = "light_weight_volume_multiplier"
+    MEDIUM_VOLUME_WEIGHT = "medium_weight_volume_multiplier"
+    HEAVY_VOLUME_WEIGHT = "heavy_weight_volume_multiplier"
+    VOLUME_WEIGH_TO_VOLUME_PERCENT = {}
+
+    LIGHT_PRICE_WEIGHT = "light_weight_price_multiplier"
+    MEDIUM_PRICE_WEIGHT = "medium_weight_price_multiplier"
+    HEAVY_PRICE_WEIGHT = "heavy_weight_price_multiplier"
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:410]: ?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 9]
```diff
@@ -244,6 +400,40 @@ def load_config(self) -> None:
...
+                element.reload_config()
+
+    def get_local_config(self):
+        return self.trading_config
+
+    @classmethod
+    def create_local_instance(cls, config, tentacles_setup_config, tentacle_config):
+        return modes_factory.create_temporary_trading_mode_with_local_config(
+            cls, config, tentacle_config
+        )
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:435]: Where is it used?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 10]
```diff
@@ -286,6 +476,18 @@ def get_required_candles_count(cls, tentacles_setup_config: tm_configuration.Ten
...
+    async def get_backtesting_plot(cls, exchange, symbol, backtesting_id, optimizer_id,
+                                   optimization_campaign, backtesting_analysis_settings):
+        try:
+            import tentacles.Meta.Keywords.scripting_library as scripting_library
+            ctx = context_management.Context.minimal(cls, logging.get_logger(cls.get_name()), exchange, symbol,
+                                                     backtesting_id, optimizer_id,
+                                                     optimization_campaign, backtesting_analysis_settings)
+            return await scripting_library.default_backtesting_analysis_script(ctx)
+        except ImportError:
+            raise ImportError("scripting_library keywords are required")
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:489]: I don't understand why a TradingMode should have a plot method

File: octobot_trading/modes/abstract_trading_mode.py [hunk 11]
```diff
@@ -354,3 +556,213 @@ async def edit_order(self, order,
...
+                    commons_tree.get_exchange_path(
+                        self.exchange_manager.exchange_name,
+                        common_enums.InitializationEventExchangeTopics.CANDLES.value
+                    ),
+                    allow_creation=False
+                ).wait(),
+                timeout
+            )
+            if not self.exchange_manager.is_future:
+                return True
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:576]: I don't think this is mandatory because the path will not exists when not in future

File: octobot_trading/modes/abstract_trading_mode.py [hunk 12]
```diff
@@ -354,3 +556,213 @@ async def edit_order(self, order,
...
+                            user_inputs,
+                            run_dbs_identifier,
+                            is_backtesting=True
+                        ))
+                        self.are_metadata_saved = True
+                except Exception:
+                    self.__class__.SAVED_RUN_METADATA_DB_BY_BOT_ID[self.exchange_manager.bot_id] = False
+                    raise
+
+    async def get_trading_metadata(self, user_inputs=None, run_dbs_identifier=None, is_backtesting=True) -> dict:
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:623]: I don't understand why this method should be placed in a trading mode

File: octobot_trading/modes/channel/abstract_mode_producer.py [hunk 13]
```diff
@@ -76,8 +81,16 @@ def __init__(self, channel, config, trading_mode, exchange_manager):
         # Define trading modes default consumer priority level
         self.priority_level: int = channel_enums.ChannelConsumerPriorityLevels.MEDIUM.value
 
-        self.run_dbs_identifier = None
         self.symbol = None
+        self.reload_config()
+        self._is_first_call = True
+
+    def reload_config(self):
```
Comment (Herklos) [inline octobot_trading/modes/channel/abstract_mode_producer.py:88]: If I understand well it should be named `on_reloaded_config` instead of `reload_config`

File: octobot_trading/modes/channel/abstract_mode_producer.py [hunk 14]
```diff
@@ -264,21 +278,36 @@ async def finalize(self, exchange_name: str,
...
-    @contextlib.contextmanager
-    def trading_mode_trigger(self):
+    @contextlib.asynccontextmanager
+    async def trading_mode_trigger(self):
         try:
+            try:
+                if self.exchange_manager.is_backtesting and self._is_first_call:
+                    # reset_exchange_init_data done at first trading mode call in backtesting
+                    await self.trading_mode.reset_exchange_init_data()
+                    self._is_first_call = False
```
Comment (Herklos) [inline octobot_trading/modes/channel/abstract_mode_producer.py:294]: I don't understand why we have to do that

File: octobot_trading/modes/abstract_trading_mode.py [hunk 15]
```diff
@@ -354,3 +556,213 @@ async def edit_order(self, order,
...
+                commons_tree.EventProvider.instance().get_or_create_event(
+                    self.exchange_manager.bot_id,
+                    commons_tree.get_exchange_path(
+                        self.exchange_manager.exchange_name,
+                        common_enums.InitializationEventExchangeTopics.CANDLES.value
+                    ),
+                    allow_creation=False
+                ).wait(),
+                timeout
+            )
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:574]: Can we create an util method that takes an exchange_manager and a topic to wait for an init event to prevent duplicating this code?

File: octobot_trading/modes/channel/abstract_mode_producer.py [hunk 16]
```diff
@@ -330,3 +359,70 @@ def all_databases(self):
...
+            await commons_configuration.clear_user_inputs(
+                databases.RunDatabasesProvider.instance().get_run_db(self.trading_mode.bot_id)
+            )
+        await self._register_and_apply_required_user_inputs(
+            self.get_context(None, None, self.trading_mode.symbol, None, None, None, None, None, True)
+        )
+
+    async def _register_and_apply_required_user_inputs(self, context):
+        if context.exchange_manager.is_future:
+            await basic_keywords.set_leverage(context, await basic_keywords.user_select_leverage(context))
```
Comment (Herklos) [inline octobot_trading/modes/channel/abstract_mode_producer.py:393]: ?

File: octobot_trading/modes/modes_factory.py [hunk 17]
```diff
@@ -17,14 +17,13 @@
...
 
 import octobot_trading.errors as errors
-import octobot_trading.modes as modes
 
 LOGGER_TAG = "TradingModeFactory"
 
 
 async def create_trading_modes(config: dict,
-                               exchange_manager: object,
-                               trading_mode_class: modes.AbstractTradingMode.__class__,
```
Comment (Herklos) [inline octobot_trading/modes/modes_factory.py:27]: ?

File: octobot_trading/modes/modes_factory.py [hunk 18]
```diff
@@ -88,6 +87,16 @@ async def create_trading_mode(trading_mode_class: modes.AbstractTradingMode.__cl
         raise e
 
 
+def create_temporary_trading_mode_with_local_config(
```
Comment (Herklos) [inline octobot_trading/modes/modes_factory.py:90]: What is the use case?

File: octobot_trading/modes/script_keywords/basic_keywords/configuration.py [hunk 19]
```diff
@@ -26,14 +27,17 @@ async def user_select_leverage(
...
-async def user_select_emit_trading_signals(ctx, identifier, def_val=False):
-    await user_inputs.user_input(ctx, constants.CONFIG_TRADING_SIGNALS_STRATEGY, "text", identifier,
-                                 show_in_summary=False, show_in_optimizer=False)
-    return await user_inputs.user_input(ctx, "emit_trading_signals", "boolean", def_val,
-                                        show_in_summary=False, show_in_optimizer=False)
+async def user_select_emit_trading_signals(ctx, identifier, def_val=False) -> bool:
+    if is_emitting_signals := await user_inputs.user_input(ctx, constants.CONFIG_EMIT_TRADING_SIGNALS,
+                                                           commons_enums.UserInputTypes.BOOLEAN.value, def_val,
+                                                           show_in_summary=False, show_in_optimizer=False):
+        await user_inputs.user_input(ctx, constants.CONFIG_TRADING_SIGNALS_STRATEGY,
```
Comment (Herklos) [inline octobot_trading/modes/script_keywords/basic_keywords/configuration.py:37]: 👍

File: octobot_trading/personal_data/orders/channel/orders_updater.py [hunk 20]
```diff
@@ -103,6 +120,19 @@ async def _open_orders_fetch_and_push(self, is_from_bot=True, limit=ORDERS_UPDAT
                                 is_from_bot=is_from_bot)
             else:
                 await self.handle_post_open_order_update(symbol, open_orders, False)
+            if not self._initialized_event:
+                self._set_initialized_event(symbol)
+        self._initialized_event = True
+
+    def _set_initialized_event(self, symbol):
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/channel/orders_updater.py:127]: 👍

File: octobot_trading/exchange_data/exchange_symbol_data.py (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_trading/exchange_data/exchange_symbol_data.py:97]: Can we ? to me these are different things

File: octobot_trading/exchanges/exchange_manager.py (hunk 2 above)
Comment (GuillaumeDSM) [reply octobot_trading/exchanges/exchange_manager.py:85]: We are already on debug. I added that because we are creating temporary exchange managers when checking user accounts (to display credentials validity and exchange support status). We are creating an exchange manager to use the unified exchange builder system (and instanciate the exchange with the right tentacle, config etc). Because of that we are creating as many exchange managers as there are exchanges in the person's config and logging exchange stop creates a ton of useless logs

File: octobot_trading/modes/abstract_trading_mode.py (hunk 3 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:65]: We need to have a way to know if we already saved the run metadata or not for a given bot. We don't want to save the same metadata as many times as there are instances of trading modes (which is one per symbol per exchange, usually)

File: octobot_trading/modes/abstract_trading_mode.py (hunk 4 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:104]: right, moving it

File: octobot_trading/modes/abstract_trading_mode.py (hunk 5 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:155]: order is the order of appearance in the config form. We could change it to `constant.LAST_INPUT-1` for example, do you think it would be better ?

File: octobot_trading/modes/abstract_trading_mode.py (hunk 8 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:410]: woops

File: octobot_trading/modes/abstract_trading_mode.py (hunk 9 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:435]: in `AbstractTentacle` when we want user inputs from a tentacle that is not running at in the current bot:
image

File: octobot_trading/modes/abstract_trading_mode.py (hunk 10 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:489]: you are right, i'm moving it

File: octobot_trading/modes/abstract_trading_mode.py (hunk 12 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:623]: right, it's the same thing as the things marked as `#TODO remove when proper run storage strategy`, we need to discuss where to but it. for now i'm moving the method to be next to the ones we need to move

File: octobot_trading/modes/channel/abstract_mode_producer.py (hunk 13 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/channel/abstract_mode_producer.py:88]: Right, I'm changing that

File: octobot_trading/modes/channel/abstract_mode_producer.py (hunk 14 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/channel/abstract_mode_producer.py:294]: In backtesting, we can't wait for init events in the same way we do in live. In live when init events are triggered, we call `reset_exchange_init_data` which saves candles and other things. We need a way to know when to call this `reset_exchange_init_data` in backtesting and we can't really rely on these event (or we need to use an independant async task that would not be managed by the backtesting engine). The cleanest way I found was this call but it's not the perfect solution. Once we have a proper storage system we will not to do this anymore as it will be the storage system that will call what is currently called by `reset_exchange_init_data` and it will do it at the right moment.
I'm adding a comment to remember to change this

File: octobot_trading/exchanges/exchange_manager.py (hunk 2 above)
Comment (Herklos) [reply octobot_trading/exchanges/exchange_manager.py:85]: I see. Thanks for the explanation.

File: octobot_trading/modes/abstract_trading_mode.py (hunk 9 above)
Comment (Herklos) [reply octobot_trading/modes/abstract_trading_mode.py:435]: Right thanks 👍

File: octobot_trading/modes/abstract_trading_mode.py (hunk 12 above)
Comment (Herklos) [reply octobot_trading/modes/abstract_trading_mode.py:623]: 👍

File: octobot_trading/modes/modes_factory.py (hunk 17 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/modes_factory.py:27]: `object` is not accurate enough, it's triggering many IDE warnings so I think it's better not to type it at all.
importing `modes.AbstractTradingMode` is not possible anymore since `AbstractTradingMode` imports the factory

File: octobot_trading/modes/modes_factory.py (hunk 18 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/modes_factory.py:90]: this https://github.com/Drakkar-Software/OctoBot-Trading/pull/738#discussion_r985087531

File: octobot_trading/exchange_data/exchange_symbol_data.py (hunk 1 above)
Comment (Herklos) [reply octobot_trading/exchange_data/exchange_symbol_data.py:97]: If I remember well it's an equivalent of the new price event.

File: octobot_trading/modes/modes_factory.py (hunk 17 above)
Comment (Herklos) [reply octobot_trading/modes/modes_factory.py:27]: Understood, thanks.

File: octobot_trading/modes/modes_factory.py (hunk 18 above)
Comment (Herklos) [reply octobot_trading/modes/modes_factory.py:90]: Yes

File: octobot_trading/modes/channel/abstract_mode_producer.py (hunk 16 above)
Comment (Herklos) [reply octobot_trading/modes/channel/abstract_mode_producer.py:393]: Sorry, my comment wasn't clear. I was wondering why this code is in "_register_and_apply_required_user_inputs". For the rest I agree with forcing the leverage value

Comment (GuillaumeDSM) [reply octobot_trading/modes/channel/abstract_mode_producer.py:393]: The idea with `_register_and_apply_required_user_inputs` is to add and apply used inputs that should always be there when running a trading mode. I think managing leverage via user inputs makes sense so I put it there. Where would you see it ?

Comment (Herklos) [reply octobot_trading/modes/channel/abstract_mode_producer.py:380]: Oh it's only to enable the user input to set leverage. My bad, I thought it was a exchange call.

Comment (GuillaumeDSM) [reply octobot_trading/modes/channel/abstract_mode_producer.py:380]: It might end up with an exchange call if the leverage of the given contract on exchange is different (by calling trader.set_leverage())

File: octobot_trading/modes/abstract_trading_mode.py (hunk 5 above)
Comment (Herklos) [reply octobot_trading/modes/abstract_trading_mode.py]: Yes it would be better😉

File: octobot_trading/modes/script_keywords/basic_keywords/user_inputs.py [hunk 21]
```diff
@@ -39,24 +40,20 @@ async def user_input(
...
     """
     Set and return a user input value.
     Types are: int, float, boolean, options, multiple-options, text
     :return:
     """
-    tentacle_type_str = configuration.get_user_input_tentacle_type(ctx.tentacle)
-    config = ctx.tentacle.trading_config if tentacle_type_str == commons_enums.UserInputTentacleTypes.TRADING_MODE.value\
-        else ctx.tentacle.specific_config
-    return await configuration.user_input(
+    created_input = _get_user_input_factory(ctx).user_input(
```
Comment (Herklos) [inline octobot_trading/modes/script_keywords/basic_keywords/user_inputs.py:51]: 👍

File: octobot_trading/modes/abstract_trading_mode.py [hunk 22]
```diff
@@ -153,12 +153,15 @@ def get_mode_consumer_classes(self) -> list:
     def should_emit_trading_signals_user_input(self, inputs: dict):
         if self.UI.user_input(
             common_constants.CONFIG_EMIT_TRADING_SIGNALS, common_enums.UserInputTypes.BOOLEAN, False, inputs,
-            title="Emit trading signals on Astrolab for people to follow.", order=998
+            title="Emit trading signals on Astrolab for people to follow.",
+            order=commons_configuration.UserInput.MAX_ORDER - 2
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:139]: 👍

File: octobot_trading/storage/candles_storage.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _get_db(self, symbol):
+        return commons_databases.RunDatabasesProvider.instance().get_symbol_db(
+            self.exchange_manager.bot_id,
+            self.exchange_manager.exchange_name,
+            symbol
+        )
+
+
+def _format_transaction(transaction, chart, x_multiplier, kind, mode, y_data):
```
Comment (Herklos) [inline octobot_trading/storage/candles_storage.py:82]: Why is there a format_transaction in `candles_storage.py`?

File: octobot_trading/storage/portfolio_storage.py [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library
+import octobot_commons.enums as commons_enums
+import octobot_commons.constants as commons_constants
+import octobot_commons.databases as commons_databases
+
+import octobot_trading.storage.abstract_storage as abstract_storage
+
+
+class PortfolioStorage(abstract_storage.AbstractStorage):
```
Comment (Herklos) [inline octobot_trading/storage/portfolio_storage.py:23]: 👍

File: octobot_trading/storage/storage_manager.py [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+            self.trades_storage or self._trade_storage_factory(),
+            self.transactions_storage or self._transaction_storage_factory(),
+            self.candles_storage or self._candles_storage_factory(),
+            self.portfolio_storage or self._portfolio_storage_factory(),
+        )
+
+    def _trade_storage_factory(self):
+        self.trades_storage = trades_storage.TradesStorage(
+            self.exchange_manager,
+            commons_display.PlotSettings(
```
Comment (Herklos) [inline octobot_trading/storage/storage_manager.py:78]: Is it mandatory to set Display settings during Storage init ? It doesn't seem related to me.

Comment (GuillaumeDSM) [reply octobot_trading/storage/storage_manager.py:78]: It's defining the way we want to display these elements and is also defining necessary plotting info like what is the x and y coordinate for an element that doesnt necessarily have it (ex: transactions). The idea is for it to be customizable during a run (via keywords that would edit the `PlotSettings` of a storage) and initialized with default values, those are the default values.

File: octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py [hunk 26]
```diff
@@ -77,11 +84,21 @@ async def on_new_value(self, timestamp, value_by_currency, force_update=False, s
         timestamp
         :return: True if something changed
         """
+        self.last_update_time = self.portfolio_manager.exchange_manager.exchange.get_exchange_current_time()
+        previous_ending_portfolio = self.ending_portfolio
+        self.ending_portfolio = portfolio_util.portfolio_to_float(
+            self.portfolio_manager.portfolio.portfolio
+        )
```
Comment (Herklos) [inline octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py:91]: Do we really want to do this every time the portfolio is updated?

File: octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py [hunk 27]
```diff
@@ -165,6 +194,14 @@ def _load_historical_values(self, dict_values):
             for element in dict_values
         })
 
+    def _load_metadata(self, metadata):
+        if metadata:
+            metadata = metadata[0]
```
Comment (Herklos) [inline octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py:199]: What's `metadata[0]`?

File: octobot_trading/exchanges/abstract_exchange.py [hunk 28]
```diff
@@ -75,7 +75,8 @@ def __init__(self, config, exchange_manager):
 
         self.is_unreachable = False
 
-        self.load_user_inputs(self.exchange_manager.tentacles_setup_config, self.tentacle_config)
+        if self.exchange_manager.tentacles_setup_config is not None:
```
Comment (Herklos) [inline octobot_trading/exchanges/abstract_exchange.py:78]: Is it possible?

File: octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py [hunk 29]
```diff
@@ -198,7 +198,7 @@ def _load_metadata(self, metadata):
         if metadata:
             metadata = metadata[0]
             # on real trader, use historical values
-            if not self.portfolio_manager.exchange_manager.trader.is_trader_simulated:
+            if not self.portfolio_manager.exchange_manager.trader.simulate:
```
Comment (Herklos) [inline octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py:201]: Or exchange_manager.is_trader_simulated

File: octobot_trading/storage/storage_manager.py [hunk 30]
```diff
@@ -56,6 +57,8 @@ async def stop(self):
             self.portfolio_storage = self.candles_storage = None
 
     async def store_history(self):
+        if self.exchange_manager is None:
+            raise commons_errors.MissingExchangeDataError("This exchange storage has been stopped")
```
Comment (Herklos) [inline octobot_trading/storage/storage_manager.py:61]: 👍

File: octobot_trading/api/trader.py [hunk 31]
```diff
@@ -76,4 +76,4 @@ async def sell_currency_for_reference_market(exchange_manager, currency) -> list
 
 
 def get_current_bot_live_id(config):
-    return util.get_current_bot_live_id(config)
\ No newline at end of file
+    return util.get_current_bot_live_id(config)
```
[the remark this answers is not in the document]
Comment (GuillaumeDSM) [reply octobot_trading/api/trader.py]: reverted

File: octobot_trading/storage/abstract_storage.py [hunk 32]
```diff
@@ -61,14 +63,39 @@ async def register_live_consumer(self):
...
+                await self.start()
+            else:
+                await self.stop(clear=False)
+
+    async def stop(self, clear=True):
         if self.consumer is not None:
             await self.consumer.stop()
-        self.consumer = None
-        self.exchange_manager = None
+        if clear:
```
[the remark this answers is not in the document]
Comment (GuillaumeDSM) [reply octobot_trading/storage/abstract_storage.py:78]: When disabling a storage

File: octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py (hunk 27 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py:199]: updated

File: octobot_trading/exchanges/abstract_exchange.py (hunk 28 above)
Comment (GuillaumeDSM) [reply octobot_trading/exchanges/abstract_exchange.py:78]: In tests it is

File: octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py (hunk 26 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py]: I changed it to be updated only on portfolio update

**Review by Herklos — APPROVED** [review]
👍
